builtin/credential/approle/path_role.go

```diff
@@ -0,0 +1,1544 @@
+			return logical.ErrorResponse("token_max_ttl should be greater than or equal to token_ttl"), nil
+		}
+		return nil, b.setRoleEntry(req.Storage, roleName, role)
+	} else {
+		return logical.ErrorResponse("missing token_max_ttl"), nil
+	}
+}
+
+func (b *backend) pathRoleTokenMaxTTLRead(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role_name").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role_name"), nil
+	}
+
+	if role, err := b.roleEntry(req.Storage, strings.ToLower(roleName)); err != nil {
+		return nil, err
+	} else if role == nil {
+		return nil, nil
+	} else {
+		role.TokenMaxTTL /= time.Second
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"token_max_ttl": role.TokenMaxTTL,
+			},
+		}, nil
+	}
+}
+
+func (b *backend) pathRoleTokenMaxTTLDelete(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role_name").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role_name"), nil
+	}
+
+	role, err := b.roleEntry(req.Storage, strings.ToLower(roleName))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	role.TokenMaxTTL = time.Second * time.Duration(data.GetDefaultOrZero("token_max_ttl").(int))
+
+	return nil, b.setRoleEntry(req.Storage, roleName, role)
+}
+
+func (b *backend) pathRoleSecretIDUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	secretID, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate SecretID:%s", err)
+	}
+	return b.handleRoleSecretIDCommon(req, data, secretID)
+}
+
+func (b *backend) pathRoleCustomSecretIDUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	return b.handleRoleSecretIDCommon(req, data, data.Get("secret_id").(string))
+}
+
+func (b *backend) handleRoleSecretIDCommon(req *logical.Request, data *framework.FieldData, secretID string) (*logical.Response, error) {
+	roleName := data.Get("role_name").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role_name"), nil
+	}
+
+	if secretID == "" {
+		return logical.ErrorResponse("missing secret_id"), nil
+	}
+
+	role, err := b.roleEntry(req.Storage, strings.ToLower(roleName))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse(fmt.Sprintf("role %s does not exist", roleName)), nil
+	}
+
+	// Currently, only one type of bound is implemented.
+	// Ensure that it is enabled.
+	if !role.BoundSecretID {
+		return logical.ErrorResponse("bound_secret_id is not set on the role"), nil
+	}
+
+	secretIDStorage := &secretIDStorageEntry{
+		SecretIDNumUses: role.SecretIDNumUses,
+		SecretIDTTL:     role.SecretIDTTL,
+	}
+
+	metadata := data.Get("metadata").(string)
+	if metadata != "" {
+		json.Unmarshal([]byte(metadata), &secretIDStorage.Metadata)
```

Comment:
I wonder if we should require a json object or a simple `<key>=<value>,<key>=<value>`. This way certainly allows more complex metadata, so maybe that's worth it.